src/PlaybackLooping.cpp

```diff
@@ -0,0 +1,208 @@
+		if (processingEvents == ProcessingEvents::kClick)
+		{
+			const auto& viewInfo = ViewInfo::Get(*pProject);
+			const auto& playRegion = viewInfo.playRegion;
+			mOldPlayRegion = playRegion;
+			//SelectUtilities::LockPlayRegion(*pProject);
+			EnableLooping(true);
+
+			auto loopingPos = viewInfo.PositionToTime(mousePosX, viewInfo.GetLeftOffset(), false);
+			mLoopedPlayRegion.SetStart(loopingPos);
+
+			mProcessingState = ProcessingStates::kEnterLooping;
+		}
+		break;
+
+	case ProcessingStates::kEnterLooping:
+		if (processingEvents == ProcessingEvents::kDrag)
+		{
+			const auto& viewInfo = ViewInfo::Get(*pProject);
+			const auto& playRegion = viewInfo.playRegion;
+
+			if (!IsLoopingEnabled() /*&& !playRegion.Locked()*/)
+				return false;
+
+			auto width = viewInfo.GetTracksUsableWidth();
+			mousePosX = std::max(mousePosX, viewInfo.GetLeftOffset());
+			mousePosX = std::min(mousePosX, viewInfo.GetLeftOffset() + width - 1);
+
+			auto loopingPos = viewInfo.PositionToTime(mousePosX, viewInfo.GetLeftOffset(), false);
+			mLoopedPlayRegion.SetEnd(loopingPos);
+		}
+		break;
+	}
+
+	return true;
+}
+
+/* PlaybackLoopingHandle implementation */
+
+PlaybackLoopingHandle::PlaybackLoopingHandle()
+{
+	;
+}
+
+UIHandle::Result PlaybackLoopingHandle::Click (const TrackPanelMouseEvent& event, AudacityProject* pProject)
+{
+	if (event.event.LeftIsDown())
+	{
+		if (!PlaybackLooping::GetInstance().LoopingProcessor(pProject, event.event.m_x, PlaybackLooping::ProcessingEvents::kClick))
+		{
+			return RefreshCode::RefreshNone;
+		}
+	}
+
+	return RefreshCode::RefreshCell;
+}
+
+UIHandle::Result PlaybackLoopingHandle::Drag (const TrackPanelMouseEvent& event, AudacityProject* pProject)
+{
+	if (event.event.LeftIsDown())
+	{
+		if (!PlaybackLooping::GetInstance().LoopingProcessor(pProject, event.event.m_x, PlaybackLooping::ProcessingEvents::kDrag))
+		{
+			return RefreshCode::RefreshNone;
+		}
+	}
+
+	return RefreshCode::RefreshCell;
+}
+
+HitTestPreview PlaybackLoopingHandle::Preview(const TrackPanelMouseState& state, AudacityProject* pProject)
+{
+	// NOTE: need add the check on enter cursor to current loop region if looping not IDLE.
+
+	auto message = XO("PlaybackLoopingHandle message");
+	auto tooltip = XO("PlaybackLoopingHandle tooltip");
+
+	static wxCursor cursorArrow{ wxCURSOR_ARROW };
+
+	return {
+	   message,
+	   &cursorArrow,
+	   tooltip,
+	};
+}
+
+UIHandle::Result PlaybackLoopingHandle::Release (const TrackPanelMouseEvent& event, AudacityProject* pProject, wxWindow* pParent)
+{
+	// Set looping region to play region or back old paly region.
+	const auto& viewInfo = ViewInfo::Get(*pProject);
+
+	const auto& selectedRegion = viewInfo.selectedRegion;
+
+	const auto& playRegion = viewInfo.playRegion;
+
+	auto start = playRegion.GetStart();
+	auto end = playRegion.GetEnd();
+
+	return RefreshCode::RefreshCell;
+}
+
+UIHandle::Result PlaybackLoopingHandle::Cancel(AudacityProject* pProject)
+{
+	return RefreshCode::RefreshCell;
```

Comment:
The purpose of a Cancel method is to implement the ESC key, which should restore all state (here, the play region) to what it was before the click.  This is a must-do.